File: threedollar-api-core/src/main/java/com/depromeet/threedollar/application/provider/upload/UploadProvider.java
package com.depromeet.threedollar.application.provider.upload;

import com.depromeet.threedollar.application.provider.upload.dto.request.UploadFileRequest;
import com.depromeet.threedollar.external.client.storage.FileStorageClient;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor
@Component
public class UploadProvider {

    private final FileStorageClient fileStorageClient;

    public String uploadFile(UploadFileRequest request) {
        request.validateAvailableUploadFile();
        String fileName = request.getFileNameWithBucketDirectory(request.getFile().getOriginalFilename());
        fileStorageClient.uploadFile(request.getFile(), fileName);
        return fileStorageClient.getFileUrl(fileName);
    }

}

File: threedollar-infrastructure/threedollar-s3/src/main/java/com/depromeet/threedollar/infrastructure/s3/provider/UploadProvider.kt
package com.depromeet.threedollar.infrastructure.s3.provider

import java.util.concurrent.CompletableFuture
import java.util.stream.Collectors
import org.springframework.stereotype.Component
import com.depromeet.threedollar.infrastructure.s3.infra.FileStorageClient
import com.depromeet.threedollar.infrastructure.s3.provider.dto.request.UploadFileRequest
import com.depromeet.threedollar.infrastructure.s3.provider.dto.response.UploadFileWithSequenceResponse

@Component
class UploadProvider(
    private val fileStorageClient: FileStorageClient,
) {

    fun uploadFile(request: UploadFileRequest): String {
        request.validateAvailableUploadFile()
        val fileName = request.getFileNameWithBucketDirectory(request.file.originalFilename)
        fileStorageClient.uploadFile(request.file, fileName)
        return fileStorageClient.getFileUrl(fileName)
    }

    fun uploadFiles(requests: List<UploadFileRequest>): List<UploadFileWithSequenceResponse> {
        val uploadFutures = requests.mapIndexed { index, request ->
            CompletableFuture.supplyAsync {
                request.validateAvailableUploadFile()
                val fileName = request.getFileNameWithBucketDirectory(request.file.originalFilename)
                fileStorageClient.uploadFile(request.file, fileName)
                UploadFileWithSequenceResponse(
                    sequence = index,
                    fileUrl = fileStorageClient.getFileUrl(fileName)
                )
            }
        }

        return CompletableFuture.allOf(*uploadFutures.toTypedArray<CompletableFuture<*>>())
            .thenApply {
                uploadFutures.stream()
                    .map { uploadFuture -> uploadFuture.join() }
                    .collect(Collectors.toList())
            }
            .join()
    }

}
